client/redfish.c

```diff
@@ -0,0 +1,3097 @@
+		dst->subsys.vendor_id = src->subsys.vendor_id;
+		dst->subsys.device_id = src->subsys.device_id;
+		return TRUE;
+	}
+	return FALSE;
+}
+
+static ni_bool_t
+ni_mchi_net_dev_pci_eq(const ni_mchi_net_dev_pci_t *p1, const ni_mchi_net_dev_pci_t *p2)
+{
+	if (p1->vendor_id != p2->vendor_id ||
+	    p1->device_id != p2->device_id ||
+	    p1->subsys.vendor_id != p2->subsys.vendor_id ||
+	    p1->subsys.device_id != p2->subsys.device_id)
+		return FALSE;
+	return TRUE;
+}
+
+static void
+ni_mchi_net_dev_pci_destroy(ni_mchi_net_dev_pci_t *pci)
+{
+	if (pci) {
+		memset(pci, 0, sizeof(*pci));
+	}
+}
+
+static ni_mchi_net_dev_t *
+ni_mchi_net_dev_new(ni_mchi_net_dev_type_t type)
+{
+	ni_mchi_net_dev_t *dev;
+
+	dev = calloc(1, sizeof(*dev));
+	if (dev) {
+		dev->refcount = 1;
+		dev->type = type;
+		ni_mchi_net_dev_usb_init(&dev->usb);
+		ni_mchi_net_dev_pci_init(&dev->pci);
+	}
+	return dev;
+}
+
+static ni_mchi_net_dev_t *
+ni_mchi_net_dev_ref(ni_mchi_net_dev_t *dev)
+{
+	if (dev) {
+		ni_assert(dev->refcount);
+		dev->refcount++;
+	}
+	return dev;
+}
+
+static void
+ni_mchi_net_dev_destroy(ni_mchi_net_dev_t *dev)
+{
+	if (dev) {
+		dev->type = 0;
+		ni_netdev_ref_destroy(&dev->device);
+		ni_mchi_net_dev_usb_destroy(&dev->usb);
+		ni_mchi_net_dev_pci_destroy(&dev->pci);
+	}
+}
+
+static void
+ni_mchi_net_dev_free(ni_mchi_net_dev_t *dev)
+{
+	if (dev) {
+		ni_assert(dev->refcount);
+		dev->refcount--;
+		if (dev->refcount == 0) {
+			ni_mchi_net_dev_destroy(dev);
+			free(dev);
+		}
+	}
+}
+
+static ni_bool_t
+ni_mchi_net_redfish_init(ni_mchi_net_redfish_t *redfish)
+{
+	if (redfish) {
+		memset(redfish, 0, sizeof(*redfish));
+		return TRUE;
+	}
+	return FALSE;
+}
+
+static void
+ni_mchi_net_redfish_destroy(ni_mchi_net_redfish_t *redfish)
+{
+	if (redfish) {
+		ni_string_free(&redfish->service.host);
+		ni_mchi_net_redfish_init(redfish);
+	}
+}
+
+static ni_mchi_net_pconf_t *
+ni_mchi_net_pconf_new(ni_mchi_net_pconf_type_t type)
+{
+	ni_mchi_net_pconf_t *pconf;
+
+	if ((pconf = calloc(1, sizeof(*pconf)))) {
+		pconf->type = type;
+		ni_mchi_net_redfish_init(&pconf->redfish);
+	}
+	return pconf;
+}
+
+static void
+ni_mchi_net_pconf_free(ni_mchi_net_pconf_t *pconf)
+{
+	if (pconf) {
+		ni_mchi_net_redfish_destroy(&pconf->redfish);
+		pconf->type = 0;
+		pconf->next = NULL;
+	}
+}
+
+static ni_bool_t
+ni_mchi_net_pconf_list_append(ni_mchi_net_pconf_t **list, ni_mchi_net_pconf_t *pconf)
+{
+	ni_mchi_net_pconf_t *item;
+
+	if (list && pconf) {
+		while ((item = *list))
+			list = &item->next;
+		*list = pconf;
+		return TRUE;
+	}
+	return FALSE;
+}
+
+static void
+ni_mchi_net_pconf_list_destroy(ni_mchi_net_pconf_t **list)
+{
+	ni_mchi_net_pconf_t *item;
+
+	if (list) {
+		while ((item = *list)) {
+			*list = item->next;
+			ni_mchi_net_pconf_free(item);
+		}
+	}
+}
+
+static void
+ni_mchi_entry_free(ni_mchi_entry_t *mchi)
+{
+	if (mchi) {
+		ni_mchi_net_dev_free(mchi->net.dev);
+		mchi->net.dev = NULL;
+		ni_mchi_net_pconf_list_destroy(&mchi->net.plist);
```

Comment:
were are missing a `free()` here:
```suggestion
		ni_mchi_net_pconf_list_destroy(&mchi->net.plist);
		free(mchi);
```

Reply:
Thanks, good catch!